File: 2023/07/c/main.c
#include "main.h"
#include "helper.h"
#include "run.h"
#include <ctype.h>
#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

const bool HAS_ALTERNATE = false;

typedef struct hands {
  char *cards;
  int values;
} Hands;

int getRank(char c) {
  switch (c) {
  case 'A':
    return 13;
  case 'K':
    return 12;
  case 'Q':
    return 11;
  case 'J':
    return 10;
  case 'T':
    return 9;
  case '9':
    return 8;
  case '8':
    return 7;
  case '7':
    return 6;
  case '6':
    return 5;
  case '5':
    return 4;
  case '4':
    return 3;
  case '3':
    return 2;
  case '2':
    return 1;
  case '1':
    return 0;
  default:
    return -1;
  }
}

int card_comparator(const void *v1, const void *v2) {
  const Hands *p1, *p2;
  p1 = *(const Hands **)v1;
  p2 = *(const Hands **)v2;
  for (int i = 0; i < 5; i++)
    if (p1->cards[i] != p2->cards[i])
      return getRank(p1->cards[i]) - getRank(p2->cards[i]);
  return 0;
}

int getType(char *str) {
  int *values = malloc(14 * sizeof(int)), i;
  for (i = 0; i < 14; i++)
    values[i] = 0;

  for (i = 0; i < 5; i++)
    values[getRank(str[i])]++;

  if (values[0]) {
    int temp = values[0];
    values[0] = 0;
    int idx = 0, max = 0;
    for (i = 0; i < 14; i++) {
      if (values[i] > max) {
        max = values[i];
        idx = i;
      }
    }
    values[idx] += temp;
  }

  int min = 5, max = 0, sz = 0;
  for (i = 0; i < 14; i++) {
    if (values[i] > 0) {
      sz++;
      if (min > values[i])
        min = values[i];
    }
    if (max < values[i])
      max = values[i];
  }
  free(values);

  if (sz == 1)
    return 6;
  if (sz == 4)
    return 1;
  if (sz == 5)
    return 0;
  if (min == 1) {
    if (max == 2)
      return 2;
    if (max == 3)
      return 3;
    if (max == 4)
      return 5;
  }
  return 4;
}

char *part1(char *input) {
  int sz, i, j, k = 0, res = 0;
  char **splt = strsplit(input, "\n", &sz);
  Hands ***groups = malloc(7 * sizeof(Hands **));
  int *group_sz = malloc(7 * sizeof(int));
  for (i = 0; i < 7; i++) {
    groups[i] = malloc(sz * sizeof(Hands *));
    group_sz[i] = 0;
  }

  for (i = 0; i < sz; i++) {
    int temp_sz;
    char **temp = strsplit(splt[i], " ", &temp_sz);
    Hands *hands = malloc(sizeof(Hands));
    hands->cards = malloc(strlen(temp[0]) * sizeof(char));
    strcpy(hands->cards, temp[0]);
    hands->values = atoi(temp[1]);

    int type = getType(hands->cards);
    groups[type][group_sz[type]++] = hands;

    free(temp[0]);
    free(temp[1]);
    free(temp);
    free(splt[i]);
  }
  free(splt);

  for (i = 0; i < 7; i++) {
    qsort(groups[i], group_sz[i], sizeof(Hands *), card_comparator);
    for (j = 0; j < group_sz[i]; j++) {
      res += groups[i][j]->values * ++k;
      free(groups[i][j]->cards);
      free(groups[i][j]);
    }
    free(groups[i]);
  }
  free(groups);
  free(group_sz);

  return numtostr(res);
}

char *part2(char *input) {
  int sz, i, j, k = 0, res = 0;
  char **splt = strsplit(input, "\n", &sz);
  Hands ***groups = malloc(7 * sizeof(Hands **));
  int *group_sz = malloc(7 * sizeof(int));
  for (i = 0; i < 7; i++) {
    groups[i] = malloc(sz * sizeof(Hands *));
    group_sz[i] = 0;
  }

  for (i = 0; i < sz; i++) {
    int temp_sz;
    char **temp = strsplit(splt[i], " ", &temp_sz);
    Hands *hands = malloc(sizeof(Hands));
    hands->cards = malloc(strlen(temp[0]) * sizeof(char));
    strreplacec(temp[0], 'J', '1');
    strcpy(hands->cards, temp[0]);
    hands->values = atoi(temp[1]);

    int type = getType(hands->cards);
    groups[type][group_sz[type]++] = hands;

    free(temp[0]);
    free(temp[1]);
    free(temp);
    free(splt[i]);
  }
  free(splt);

  for (i = 0; i < 7; i++) {
    qsort(groups[i], group_sz[i], sizeof(Hands *), card_comparator);
    for (j = 0; j < group_sz[i]; j++) {
      res += groups[i][j]->values * ++k;
      free(groups[i][j]->cards);
      free(groups[i][j]);
    }
    free(groups[i]);
  }
  free(groups);
  free(group_sz);

  return numtostr(res);
}

int main(int argc, char *argv[]) {
  return run(argc, argv, part1, part2, HAS_ALTERNATE);
}

File: 2023/07/ocaml/main.ml
let options = Hashtbl.create 128;;

Hashtbl.add options "has_alternate" false;;
Hashtbl.add options "has_io" false

let ( +/ ) = Int64.add
let ( */ ) = Int64.mul

let get_ranking c =
  match c with
  | 'A' -> 13
  | 'K' -> 12
  | 'Q' -> 11
  | 'J' -> 10
  | 'T' -> 9
  | '9' -> 8
  | '8' -> 7
  | '7' -> 6
  | '6' -> 5
  | '5' -> 4
  | '4' -> 3
  | '3' -> 2
  | '2' -> 1
  | '1' -> 0
  | _ -> 0
;;

let high_card = 0
let one_pair = 1
let two_pair = 2
let three_of_a_kind = 3
let full_house = 4
let four_of_a_kind = 5
let five_of_a_kind = 6

let get_type (cards : string) : int =
  let values = Array.make 14 0 in
  for i = 0 to 4 do
    values.(get_ranking cards.[i]) <- values.(get_ranking cards.[i]) + 1
  done;
  let _ =
    if values.(0) > 0
    then (
      let temp = values.(0) in
      values.(0) <- 0;
      let mx = values |> Array.fold_left max 0 in
      let idx = values |> Array.find_index (fun v -> v = mx) in
      if Option.is_some idx
      then values.(Option.get idx) <- values.(Option.get idx) + temp
      else ())
    else ()
  in
  let new_values = values |> Array.to_list |> List.filter (fun v -> v > 0) in
  if List.length new_values = 1
  then five_of_a_kind
  else if List.length new_values = 4
  then one_pair
  else if List.length new_values = 5
  then high_card
  else (
    let mn = List.fold_left min 5 new_values in
    let mx = List.fold_left max 0 new_values in
    if mn = 1
    then if mx = 2 then two_pair else if mx = 3 then three_of_a_kind else four_of_a_kind
    else full_house)
;;

let parse_input (input : string) (joker : bool) =
  let cards =
    input
    |> String.split_on_char '\n'
    |> List.map (fun line ->
      let line = String.split_on_char ' ' line |> Array.of_list in
      let cards =
        if joker
        then
          line.(0)
          |> String.to_seq
          |> Seq.map (fun c -> if c = 'J' then '1' else c)
          |> String.of_seq
        else line.(0)
      in
      cards, Int64.of_string line.(1))
  in
  let groups = Array.make 7 [] in
  cards
  |> List.iter (fun (cards, value) ->
    let idx = get_type cards in
    groups.(idx) <- (cards, value) :: groups.(idx));
  groups
;;

let solve (input : string) (joker : bool) : string =
  let groups = parse_input input joker in
  let rec loop
    (i : int)
    (groups : (string * int64) list array)
    (count : int64)
    (res : int64)
    : int64
    =
    if i = 7
    then res
    else (
      let g =
        groups.(i)
        |> List.sort (fun (c1, _) (c2, _) ->
          let res = ref 0 in
          for j = 0 to 4 do
            let x = get_ranking (String.get c1 j) in
            let y = get_ranking (String.get c2 j) in
            if x <> y && !res = 0 then res := x - y else ()
          done;
          !res)
      in
      let rec loop2 (x : int) (count : int64) (res : int64) =
        if List.length g = x
        then count, res
        else loop2 (x + 1) (count +/ 1L) (res +/ ((List.nth g x |> snd) */ count))
      in
      let new_count, new_res = loop2 0 count res in
      loop (i + 1) groups new_count new_res)
  in
  loop 0 groups 1L 0L |> Int64.to_string
;;

let part_1 (input : string) (_is_test : bool) : string = solve input false
let part_2 (input : string) (_is_test : bool) : string = solve input true
let () = Aoc.run Sys.argv part_1 part_2 options
